teps/0026-interceptor-plugins.md

```diff
@@ -0,0 +1,510 @@
+
+
+### Performance
+
+With this proposal, the number of HTTP roundtrips scales linearly with the number of interceptors. The path based Trigger approach laid out in [TEP-009](./0009-trigger-crd.md) should help reduce the number of calls since the EventListener will only process a single or a subset of all Triggers instead of every Trigger associated with an EventListener. In addition, Interceptors can be horizontally scaled so that they can handle more requests. Further, there are some low hanging fruits with respect to performance (tektoncd/triggers#797 and tektoncd/triggers#594) that we can tackle to improve performance of the EventListener. Finally, there has been a alternative proposal to chain multiple interceptors into a single "meta interceptor" to reduce HTTP calls (https://github.com/tektoncd/community/pull/229#discussion_r504698565).
+
+
+## Design Details
+
+This section describes the two major portions of the proposal -- the changes to
+the Interceptor interface to make it more typed, and the new CRD to
+install/create interceptors.
+
+### HTTP Interface
+
+This diverges from the current HTTP interface that simply forwards the given request and uses HTTP response codes. Instead the proposed way uses typed response bodies for both requests and response (similar to the AdmissionReviewRequest and Response in AdmissionControllers)
+
+
+#### Request
+
+EL Sink sends incoming events to interceptors over HTTP using a typed format:
+
+```go
+type InterceptorRequest struct {  
+	// Body is the incoming HTTP event body
+	Body []byte `json:"body,omitempty"`
+	// Header are the headers for the incoming HTTP event
+	Header map[string][]string `json:"header,omitempty"`
+
+	// InterceptorParams are the user specified params for the interceptor
+	InterceptorParams map[string]string `json:"interceptor_params,omitempty"`
+
+	Context *TriggerContext
+}
+
+type TriggerContext struct {
+	// EventUrl is the URL of the incoming event
+	EventUrl string `json:"url,omitempty"`
+	// EventID is a unique ID assigned by Triggers to each event
+	EventID string `json:"event_id,omitempty"`
+	// TriggerID is of the form namespace/$ns/triggers/$name
+	TriggerID string `json:"trigger_id,omitempty"`
+}
+```
+
+#### Response
+
+Interceptors in turn return a message with a fixed type:
+
+```go
+type InterceptorResponse struct {
+	// Extensions are  additional fields that is added to the interceptor event.
+	// See TEP-0022. Naming TBD.
+	Extensions map[string]interface{} `json:"extensions,omitempty"`
+	// Continue indicates if the EventListener should continue processing the Trigger or not
+	Continue bool `json:"continue,omitempty"`
+	// Status is an Error status t
+	Status grpc.Status `json:"status,omitempty"`
```

Comment:
might be worth linking to something with more details of what `grpc.Status` contains - what's involved for interceptor authors to create this?

Reply:
Added links to https://google.aip.dev/193